Pg queries

Danielxburns opened the pull request:
Queries users table of 10M records for user data based on userId. Tested times ranged from 1-69ms.
Queries reviews table of 100M record for all reviews based on userId. Tested once, resulted query time was 175835ms

Danielxburns committed:
generate a csv file of 1M records in 3 minutes

Danielxburns committed:
generate a csv file of 10M users

Danielxburns committed:
generate a csv file of 100M reviews, 24.G in 27 mins

Danielxburns committed:
create schema

Danielxburns committed:
Update seed scripts

-Fix bug with only generating 1M userIds instead of 10M.
-Add a line to write the headers to the csv files.
-Add 'likes' property to reviews
-Changes review property to a sentence instead of a paragraph to conserve space.

Danielxburns committed:
Connect server to database and perform basic queries

-get test info from a small test table
-get all user info by userId from users table of 10M records
-queries reviews table of 100M records for all reviews from a user based on userId

Danielxburns committed:
Connect server to DB and add basic query functions

-query testusers table of 10 records to verify server code is working
-query users table of 10M records for all data on user based on userId
-query reviews table of 100M records for all reviews based on userId

hunterxdong commented:
Good use of drain. How does this style of data generation affect ram usage?

Danielxburns commented:
RAM usage is not an issue with using drain event. The drain event has to do with the maxing out the internal buffer of the writeable stream. Here's a stackoverflow that explains a lot about it https://stackoverflow.com/questions/18932488/how-to-use-drain-event-of-stream-writable-in-node-js
I saw that you were using .pipe in your code. That might be where you're maxing out your RAM. not sure though.
I blogged about that drain event if you want to read it 
https://medium.com/@danielburnsart/writing-a-large-amount-of-data-to-a-csv-file-using-nodes-drain-event-99dcaded99b5

Danielxburns commented:
Queries users table of 10M records for user data based on userId. Tested times ranged from 1-69ms.
Queries reviews table of 100M record for all reviews based on userId. Tested once, resulted query time was 175835ms

jordisman commented:
This is interesting "writer.once('drain', write);" I didn't know there is a 'drain' option

therobinkim commented on db/index.js:
```diff
@@ -1,113 +1,80 @@
+const pool = new Pool({
+  user: 'daniel',
+  host: 'localhost',
+  database: 'bookshelf',
```
you may want to consider extracting your db credentials into a separate gitignored config file or as env variables

Danielxburns replied:
most definitely. 
done!

therobinkim commented on db/index.js:
```diff
@@ -1,113 +1,80 @@
+      console.log(error);
+    }
+    console.log(`total query time was ${Date.now() - time}ms`);
+    response.status(200).json(results.rows);
```
ending the response with a 200 status code when you get an error seems a little wonky

Danielxburns replied:
I see what you're saying. As it is now the interpreter would not exit the function after logging the error.  I'll try throwing the error. I suppose I could also add an else statement.
Follow-up: I just read a good post about error throwing explaining, among other things, how `throw new Error(error)` has broader browser support than `throw(error)`. 
https://humanwhocodes.com/blog/2009/03/10/the-art-of-throwing-javascript-errors-part-2/

therobinkim commented on server/app.js:
```diff
@@ -1,70 +1,29 @@
-    res.json(err);
-  }
+app.get('/', (request, response) => {
+  response.json({ info: 'Node.js, Express, and Postgres API' });
```
is this endpoint necessary?

Danielxburns replied:
nope. That was just there let me know I was connected when testing the server in Postman.
Deleted.

therobinkim commented on server/app.js:
```diff
@@ -1,70 +1,29 @@
 const bodyParser = require('body-parser');
-const path = require('path');
-const db = require('../db');
+const db = require('../db/index.js');
```
you can just leave this as `require('../db')`

Danielxburns replied:
nice. thanks!

therobinkim commented on server/app.js:
```diff
@@ -1,70 +1,29 @@
-  }
-});
+// GET USER INFO FROM USERS TABLE BY ID
+app.get('/users/:id', db.getUserById);
```
nice abstractions!

Danielxburns committed:
Update changes reccomended by Robin's code review

-extract db credentials to ingnored config file
-update error handling to throw instead of log
-update query timing functions to console.time

jordisman reviewed (commented):
You mentioned the comma in your data gave you trouble when generating csv, does using faker sentence solve it? How about the comma in the dates?
